### PR TITLE
Initial work for integrating the UI tests with Jenkins

### PR DESCRIPTION
**Pull request by izaac**
### Summary

This is for executing the UI tests in Jenkins UI

### Technical notes summary

The logic starts in `init.sh` which prepares the Jenkins executor with the software dependencies and an external node for test execution using [a corral package.](https://github.com/rancherlabs/corral-packages/pull/17). The execution logic is then done by the corral script executed. See the package PR.

The variables values come from the Jenkins job. That's a separate PR in a private repo. The Jenkinsfile is also included in this PR. It gets pulled by the Jenkins pipeline.

This makes use of the official Cypress' [Docker images](https://github.com/cypress-io/cypress-docker-images). The included one. The image version and type is included as variables in Jenkins.

Cypress reporter modules is installed in the Cypress docker image and the report generates jUnit and HTML reports. `reporter-options.json` is used to configure the multi-reporter.
HTML publishing in Jenkins is still pending but jUnit is ready.
`scpget.sh` is a helper script to retrieve the Cypress results from the external test node.

### CODE REVIEW

File: cypress/jenkins/cypress.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+set -x
+set -e
+
+pwd
+ls -al
+cd "dashboard"
+
+yarn install
+yarn add --dev -W mocha mochawesome mochawesome-merge \
```
Comment (richard-cox) [inline cypress/jenkins/cypress.sh:11]: comment only. how come the switch to mocha was needed?

File: cypress/jenkins/cypress.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  mochawesome-report-generator cypress-multi-reporters \
+  mocha-junit-reporter 
+
+NO_COLOR=1 cypress run --browser chrome \
+  --reporter cypress-multi-reporters \
+  --reporter-options \
+    configFile=cypress/jenkins/reporter-options.json \
+  --env grepTags=CYPRESSTAGS
```
Comment (richard-cox) [inline cypress/jenkins/cypress.sh:19]: think this needs to be `GREP_TAGS`. https://github.com/rancher/dashboard/blob/master/cypress.config.ts#L99. 

Tests now run as either an admin or a standard user. This env will ensure tests designed to run only against admin or only against standard user run. If standard user is selected you need to also supply `TEST_USERNAME` as well. @yonasberhe23 can walk you through those requirements

File: cypress/jenkins/cypress.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+pwd
+ls -al
+cd "dashboard"
+
+yarn install
+yarn add --dev -W mocha mochawesome mochawesome-merge \
+  mochawesome-report-generator cypress-multi-reporters \
+  mocha-junit-reporter 
+
+NO_COLOR=1 cypress run --browser chrome \
```
Comment (richard-cox) [inline cypress/jenkins/cypress.sh:15]: There's a few args missing here

`TEST_BASE_URL` - points to the Rancher instance tests will run on
`TEST_USERNAME` (defaults to admin) - user to run the test as
`TEST_PASSWORD` - password of the user to run the tests as.
Depending on the Rancher install you also might need

`CATTLE_BOOTSTRAP_PASSWORD` - bootstrap password that will be used to setup rancher

**Review by richard-cox — CHANGES_REQUESTED** [review]

File: cypress/jenkins/cypress.sh (hunk 1 above)
Comment (izaac) [reply cypress/jenkins/cypress.sh:11]: It seems that Cypress makes use of mocha libraries as a backend for a few things like reporting. I needed to include these  during execution to generate jUnit and html reports and merge them for Jenkins publihsing.

File: cypress/jenkins/cypress.sh (hunk 2 above)
Comment (izaac) [reply cypress/jenkins/cypress.sh:19]: CYPRESSTAGS is just a placeholder to dinamically replace it by the `GREP_TAGS` values.

All those values are read and supplied in the Corral Package `run.sh`
https://github.com/rancherlabs/corral-packages/pull/17/files#diff-0691477a86647ab7e92bdc8f54ff53460a79660fbacdbf2f5fc175ef09c97cc4

File: cypress/jenkins/cypress.sh (hunk 1 above)
Comment (richard-cox) [reply cypress/jenkins/cypress.sh:11]: cypress should dump out some reasonable html reports, think probably junit as well. though i don't think that should block this PR, and you can tailor the output without having to worry about breaking github workflows

**Review by richard-cox — APPROVED** [review]
Would it be possible at some point, possibly in a README in `cypress/jenkins`, detailing the process for how these run? Doesn't need to be long, just enough to kick-start anyone else that looks at these files

**Comment by izaac** [comment]
@richard-cox absolutely I will prepare a `README.md` for the next iteration (recurring execution) of this. 
There are many moving parts and that'd be really helpful.

**Pull request merged**